src/json-observe.ts

```diff
@@ -0,0 +1,297 @@
+}
+
+//3x faster than cached /^\d+$/.test(str)
+function isInteger(str: string): boolean {
+  var i = 0;
+  var len = str.length;
+  var charCode;
+  while (i < len) {
+    charCode = str.charCodeAt(i);
+    if (charCode >= 48 && charCode <= 57) {
+      i++;
+      continue;
+    }
+    return false; 
+  }
+  return true;
+}
+function deepClone(obj: any) {
+  switch (typeof obj) {
+    case "object":
+      return JSON.parse(JSON.stringify(obj)); //Faster than ES5 clone - http://jsperf.com/deep-cloning-of-objects/5
+
+    case "undefined":
+      return null; //this is how JSON.stringify behaves for array items
+
+    default:
+      return obj; //no need to clone primitives
+  }
+}
+/** Class representing a JS Object observer . */
+class JsonObserver {
+
+  private originalObject;
+  private exposedObject: Object;
+  private cachedProxy: Object;
+  private patches: Array<Object>;
+  private isRecording: boolean;
+  private defaultCallback: Function;
+  public userCallback: Function;
+
+  /**
+   * Disables and enables operation recording and/or callback firing when object modifications happen.
+   * @param {boolean} enable - true will enable, false will disable.
+   */
+  public disableEnableCallback: Function;
+
+  private escapePathComponent(str) {
+    if (str.indexOf('/') === -1 && str.indexOf('~') === -1) return str;
+    return str.replace(/~/g, '~0').replace(/\//g, '~1');
+  }
+  private generateProxyAtPath(obj, path) {
+
+    if (!obj) return obj;
+
+    path = path === '/' ? '' : path; //root
+    var instance = this;
+    let proxy = new Proxy(obj,
+    {
+        get: (target, propKey, receiver) => {
+          if (propKey.toString() === '_proxy') return true; //to distinguish proxies
+          return Reflect.get(target, propKey, receiver);
+        },
+        set: function (target, key, receiver) {
+
+          /* 
+          https://github.com/Starcounter-Jack/JSON-Patch/issues/125
+          TODO: I didn't remove it to benchmark it later
+          if(typeof receiver === 'function')
+          {            
+            return Reflect.set(target, key, receiver);
+          }
+          */
+
+          var distPath = path + '/' + instance.escapePathComponent(key.toString());
+          // if the new value is an object, make sure to watch it          
+          if (receiver /* because `null` is in object */ && typeof receiver === 'object' && receiver._proxy !== true) {
+            receiver = instance.generateProxyAtPath(receiver, distPath);
+          }
+
+          if (typeof receiver === 'undefined') {
+            if (target.hasOwnProperty(key))  //exists
+            {
+              // when array element is set to `undefined`, should generate replace to `null` 
+              if (_isArray(target)) {
+                //undefined array elements are JSON.stringified to `null`
+                instance.defaultCallback({ op: 'replace', path: distPath, value: null });
+                return Reflect.set(target, key, receiver);
```

Comment:
Please DRY `return Reflect.set(target, key, receiver);` you are repeating same code in all complicated if/elses